feat: added support for custom request config

amgadserry opened the pull request:
feat: added support for custom http client

# Pull Request

## Related issue


## What does this PR do?
- adds support for usages with preconfigured http clients like axios for using ms through a proxy for example
```
const client: MeiliSearch = new MeiliSearch({
      host: 'http://localhost:3000/api/v0/meilisearch/proxy',
      httpClient: async (url, opts) => {
        const response = await $charles.http!.getClient().request({
          url,
          data: opts?.body,
          headers: opts?.headers
        })

        return response.data
      }
    })
```
- adds support for customizing the requests created by `httpRequests` 

### BREAKING CHANGES:
- removes `headers` config param and introduces `requestConfig` instead which can be used to customize all fetch config

## PR checklist
Please check if your PR fulfills the following requirements:
- [x] Does this PR fix an existing issue, or have you listed the changes applied in the PR description (and why they are needed)?
- [x] Have you read the contributing guidelines?
- [x] Have you made sure that the title is accurate and descriptive of the changes?

Thank you so much for contributing to Meilisearch!

bidoubiwa reviewed (commented):
🙏 Thanks so much for this MUCH needed feature.

Could you add a test that uses another httpClient? Like axios for example

amgadserry commented on src/http-requests.ts:
```diff
@@ -40,19 +40,29 @@ function constructHostURL(host: string): string {
+function hasHeader(headers: HeadersInit, headerName: string) {
+  if (Array.isArray(headers)) {
+    return headers.some((header) => header[0] === headerName)
+  } else if ('has' in headers) {
+    return (headers as Headers).has(headerName)
+  }
```
yes Many of the http libs including `fetch` allow the usage of maps and their own implementation of header u can check the Typing of fetch config

amgadserry commented on src/http-requests.ts:
```diff
@@ -111,12 +125,20 @@ class HttpRequests {
     try {
-      const response: any = await fetch(constructURL.toString(), {
+      const fetchFn = this.httpClient ? this.httpClient : fetch
+      const _request = fetchFn(constructURL.toString(), {
```
`const client: MeiliSearch = new MeiliSearch({
      host: 'http://localhost:3000/api/v0/meilisearch/proxy',
      httpClient: async (url, opts) => {
        const response = await $http!.getClient().request({
          url,
          data: opts?.body,
          headers: opts?.headers,
          method: (opts?.method?.toLocaleUpperCase() as Method) ?? 'GET'
        })

        return response.data
      }
    }))`
    
    in case the user provided an httpClient such as the example above

amgadserry replied:
I will add tests and examples I just wanted to make sure this is something u r open to merge first

haayhappen commented on src/http-requests.ts:
```diff
@@ -40,19 +40,29 @@ function constructHostURL(host: string): string {
+    return headers.some((header) => header[0] === headerName)
+  } else if ('has' in headers) {
+    return (headers as Headers).has(headerName)
+  }
+
+  return !!headers[headerName]
+}
+
 function createHeaders(config: Config): Record<string, any> {
   const agentHeader = 'X-Meilisearch-Client'
   const packageAgent = `Meilisearch JavaScript (v${PACKAGE_VERSION})`
   const contentType = 'Content-Type'
-  config.headers = config.headers || {}
+  const configHeaders = config.requestConfig?.headers ?? {}
```
Don't we need a fallback to config.headers for backwards compatibility?

haayhappen reviewed (approved):
LGTM, just one question

bidoubiwa commented:
Thanks for the response @amgadserry. We definitely want to merge this PR :)

bidoubiwa commented:
Are you able to reproduce this failing test locally?

amgadserry commented:
ah ty for the explanation 

I rebased 

but no everything is fine locally

amgadserry commented:
@bidoubiwa can we run this on main and validate that it's PR related first ?

amgadserry commented:
Fixed: It seems fetch on browser has to stay bound to window object and assigning it in constructor bound it to the class instance instead so I reverted that part of the code and everything seemed to work 
https://fetch.spec.whatwg.org/#concept-request-window

bidoubiwa commented:
> might want to add yarn build &&  at the beginning of test:env:browser command just to be safe

Fixed #1467

bidoubiwa commented on README.md:
```diff
@@ -327,6 +327,46 @@ index
+    host: 'http://localhost:3000/api/meilisearch/proxy',
+    requestConfig: {
+        headers: {
+            Authorization: AUTH_TOKEN
```
```suggestion
            Authorization: `Bearer ${AUTH_TOKEN}`
```

WDYT ?

amgadserry replied:
doesn't have to be a bearer token really this should be custom per user needs as it's just an example

bidoubiwa commented on README.md:
```diff
@@ -327,6 +327,46 @@ index
+
+### Using Meilisearch behind a proxy <!-- omit in toc -->
+
+#### Custom request config <!-- omit in toc -->
```
In both `Custom request config` and `Custom http client` can you change the indentation to 2 spaces. Some are 4 spaces and some are missing indentation..

bidoubiwa commented on src/http-requests.ts:
```diff
@@ -111,12 +130,23 @@ class HttpRequests {
 
     try {
-      const response: any = await fetch(constructURL.toString(), {
+      const _fetchFn = this.httpClient ? this.httpClient : fetch
+      const _request = _fetchFn(constructURL.toString(), {
```
What is the convention to add the underscore before the variable name?

bidoubiwa commented on README.md:
```diff
@@ -327,6 +327,46 @@ index
+
+#### Custom http client <!-- omit in toc -->
+
+You can use your own http client ex: axios
```
```suggestion
You can use your own HTTP client, for example, with [`axios`](https://github.com/axios/axios).
```

bidoubiwa commented on README.md:
````diff
@@ -327,6 +327,46 @@ index
+
+```ts
+const client: MeiliSearch = new MeiliSearch({
+    host: 'http://localhost:3000/api/meilisearch/proxy',
````
```suggestion
    host: 'http://localhost:7070/api/meilisearch/proxy',
```

as per the port we used in the other examples

amgadserry replied:
usually the proxy wouldn't be on the same port that's the whole idea

bidoubiwa commented on README.md:
````diff
@@ -327,6 +327,46 @@ index
+
+```ts
+const client: MeiliSearch = new MeiliSearch({
+    host: 'http://localhost:3000/api/meilisearch/proxy',
````
```suggestion
    host: 'http://localhost:7700/api/meilisearch/proxy',
```

bidoubiwa commented on src/http-requests.ts:
```diff
@@ -111,12 +130,23 @@ class HttpRequests {
-      }).then((res) => httpResponseErrorHandler(res))
+      })
+
+      // If this is done using custom http client return response as is to allow the user to parse/handle it as they see fit
```
```suggestion
      // When using a custom HTTP client, the response is returned to allow the user to parse/handle it as they see fit
```

bidoubiwa commented on README.md:
```diff
@@ -327,6 +327,46 @@ index
+
+#### Custom request config <!-- omit in toc -->
+
+You can customize the request generated ex: add headers
```
```suggestion
You can provide a custom request configuration. for example, with custom headers.
```

bidoubiwa commented on src/http-requests.ts:
```diff
@@ -40,19 +40,34 @@ function constructHostURL(host: string): string {
   }
 }
 
+function cloneAndParseHeaders(headers: HeadersInit): Record<string, string> {
+  if (Array.isArray(headers)) {
+    return headers.reduce((acc, headerPair) => {
+      acc[headerPair[0]] = headerPair[1]
+      return acc
+    }, {} as Record<string, string>)
+  } else if ('has' in headers) {
+    const clonedHeaders: Record<string, string> = {}
+    ;(headers as Headers).forEach((value, key) => (clonedHeaders[key] = value))
+    return clonedHeaders
+  } else {
+    return Object.assign({}, headers)
+  }
+}
```
Sorry I did not ask this sooner.

Could you add two tests, one with headers as a Map and one as a nested array?

Example for nested array:

```ts
  const client = new MeiliSearch({
        ...config,
        apiKey: '',
        requestConfig: {
          headers: [['Authorization', `Bearer ${key}`]], 
        },
      })
```

example with Map()
```ts
      const headers = new Map()
      headers.set('Authorization', `Bearer ${key}`)

      const client = new MeiliSearch({
        ...config,
        apiKey: '',
        requestConfig: {
          // @ts-ignore
          headers,
        },
      })
```

As you can see I added a `ts-ignore` on the Map test.  It seems that [`HeadersInit`](https://microsoft.github.io/PowerBI-JavaScript/modules/_node_modules_typedoc_node_modules_typescript_lib_lib_dom_d_.html#headersinit) does not accept `Map`. 

If you think it is a common use case we can keep it and keep the ignore comment, but I'm very curious if you have any example of a lib or a use-case where Map is used.

amgadserry replied:
sure 

it's not a map it accepts Headers object just use new Headers() instead will add tests

amgadserry replied:
added tests

bidoubiwa reviewed (approved):
LGTM 🔥 Thanks a lot

bidoubiwa commented:
bors merge
